Enhance zos outputs

jcarpanelli opened the pull request:
This PR implements several improvements to `zos` outputs. Some of them are:
- Improvements in log messages texts.
- Hide unnecessary logs on non-verbose mode.
- Implement multispinners using [spinnies](https://github.com/jcarpanelli/spinnies), a super handy and fancy lib for managing multiple spinners in command-line interface programs :stuck_out_tongue:.
- Reimplemented the `verbose` mode, adding a timestamp plus the name of the function/method where the log is being printed

**Bonus track**:

I removed the `StatusChecker`, `StatusComparator` and `StatusFetcher` together with the `pull`, `fetch` and `status` scripts. I also added an error message in the `status` command notifying that it's deprecated.

**TODO**
- ~I removed all references to the old `Logger` methods (warn, info, error). I still have to remove all the unused imports, the `Logger.silent(true)` in some config files and also have to refactor a bit the `unpack` command to work directly with the new logger instead of using the `spinner` utils.~
- ~Still have to finish fixing the `CaptureLogs` class~
- ~Remove `stdouts` that we discussed offline.~


A mini demo of how its working (running `zos create` before pushing contracts + pushing an EVM-package locally)

Before:
![before](https://i.ibb.co/DRjtX9F/old-outputs.gif)

After:
![after](https://i.ibb.co/5kF3cZ5/new-outputs.gif)

spalladino commented on packages/cli/src/scripts/remove.ts:
```diff
@@ -10,5 +15,11 @@ export default function remove({
+  Loggy.add(
+    `${fileName}#remove`,
+    'remove-contracts',
+    `All specified contracts have been successfully removed from the project. To add them again, just run 'zos add'`,
```
```suggestion
    `All specified contracts have been removed from the project. To add them again, just run 'zos add'.`,
```

spalladino commented on packages/lib/src/application/App.ts:
```diff
@@ -28,13 +30,17 @@ export default class App {
   }
 
   public static async deploy(txParams: TxParams = {}): Promise<App> {
-    log.info('Deploying new App...');
```
Should we add this as a Loggy.add?

jcarpanelli replied:
Moved it to `AppProject.ts`

spalladino commented on packages/lib/src/application/App.ts:
```diff
@@ -151,33 +134,45 @@ export default class App {
           );
-    log.info(`${packageName} ${contractName} proxy: ${proxy.address}`);
+    Loggy.succeed(
+      `create-proxy-${implementation}`,
```
Are we creating more than a single proxy at the same time? If so, why do we need to use the implementation as id and retrieve it earlier?

spalladino commented on packages/lib/src/application/ImplementationDirectory.ts:
```diff
@@ -15,10 +17,12 @@ export default class ImplementationDirectory {
     txParams: TxParams = {},
   ): Promise<ImplementationDirectory> {
     const contract = this.getContract();
-    log.info(`Deploying new ${contract.schema.contractName}...`);
```
Should we add this deploying in Loggy?

spalladino commented on packages/lib/src/application/ImplementationDirectory.ts:
```diff
@@ -68,35 +72,56 @@ export default class ImplementationDirectory {
-    log.info('Frozen');
+    Loggy.succeed(
+      `freeze-implementation`,
+      `Directory version successfully freezed`,
```
```suggestion
      `Directory version frozen`,
```

spalladino commented on packages/lib/src/application/Package.ts:
```diff
@@ -24,14 +26,18 @@ export default class Package {
   }
 
   public static async deploy(txParams: TxParams = {}): Promise<Package> {
-    log.info('Deploying new Package...');
```
Should this be a `Loggy.add("deploying")`?

jcarpanelli replied:
Same as with App, I added these messages in `AppProject.js`. The rationale behind this is that I didn't want to log these messages while pushing an external EVM-package, but just when you actually publish your project.

spalladino commented on packages/lib/src/artifacts/ZWeb3.ts:
```diff
@@ -97,8 +106,11 @@ export default class ZWeb3 {
+      Loggy.add(
+        `${fileName}#toChecksumAddress`,
+        'checksum-addresses',
         `WARNING: Address ${address} is not checksummed. Consider checksumming it to avoid future warnings or errors.`,
```
Let's make a note to improve this message...

spalladino commented on packages/lib/src/project/AppProject.ts:
```diff
@@ -63,18 +65,53 @@ class BaseAppProject extends BasePackageProject {
+        );
+        thepackage = await Package.deploy(txParams);
+        Loggy.update(`publish-project`, {
+          text: 'Deploying new Package contract',
```
Isn't this updating the spinner with the same text as before?

jcarpanelli replied:
Good catch!

spalladino commented on packages/lib/src/project/AppProject.ts:
```diff
@@ -63,18 +65,53 @@ class BaseAppProject extends BasePackageProject {
+        directory = await thepackage.newVersion(version);
+        const succeedText =
+          !appAddress || !packageAddress
+            ? `Project structure successfully deployed!`
```
```suggestion
            ? `Project structure deployed!`
```

spalladino commented on packages/lib/src/project/AppProject.ts:
```diff
@@ -63,18 +65,53 @@ class BaseAppProject extends BasePackageProject {
+        const succeedText =
+          !appAddress || !packageAddress
+            ? `Project structure successfully deployed!`
+            : `Version ${version} successfully added`;
```
```suggestion
            : `Version ${version} deployed`;
```

spalladino commented on packages/lib/src/proxy/ProxyAdmin.ts:
```diff
@@ -22,13 +22,20 @@ export default class ProxyAdmin {
-    log.info(`Deployed ProxyAdmin at ${contract.address}`);
+    Loggy.succeed(
+      `deploy-proxy-admin`,
+      `Successfully deployed ProxyAdmin at ${contract.address}`,
```
```suggestion
      `Deployed ProxyAdmin at ${contract.address}`,
```

spalladino commented on packages/lib/src/proxy/ProxyAdmin.ts:
```diff
@@ -75,19 +89,31 @@ export default class ProxyAdmin {
+    Loggy.add(
+      `${fileName}#upgradeProxy`,
+      `upgrade-proxy-${proxyAddress}`,
+      `TX receipt received: ${receipt.transactionHash}`,
```
```suggestion
      `Transaction receipt received: ${receipt.transactionHash}`,
```

spalladino commented on packages/lib/src/proxy/ProxyAdmin.ts:
```diff
@@ -75,19 +89,31 @@ export default class ProxyAdmin {
-    log.info(`Owner for proxy admin set to ${newAdminOwner}`);
+    Loggy.succeed(
+      'transfer-ownership',
+      `Owner for proxy admin set to ${newAdminOwner}`,
```
```suggestion
      `Proxy admin owner set to ${newAdminOwner}`,
```

spalladino commented on packages/lib/src/proxy/ProxyAdmin.ts:
```diff
@@ -75,19 +89,31 @@ export default class ProxyAdmin {
+    Loggy.add(
+      `${fileName}#transferOwnership`,
+      'transfer-ownership',
+      `Changing ownership for proxy admin to ${newAdminOwner}`,
```
```suggestion
      `Changing ownership of proxy admin to ${newAdminOwner}`,
```

jcarpanelli commented:
Thanks for the review @spalladino :muscle:. I know this was a tough to review because of its length, sorry :cry:. I will work today to address most of your suggestions and will ping you when it's ready.

jcarpanelli commented on packages/cli/src/commands/status.ts:
```diff
@@ -25,13 +20,7 @@ const register: (program: any) => any = program =>
-  if (options.fix) await pull({ network, txParams });
-  else if (options.fetch) await compare({ network, txParams });
-  else await status({ network, txParams });
+  throw Error('Status command has been deprecated.');
```
I believe we can! Let me see if it's not very difficult though. If it's not, I'll include it in this PR.

jcarpanelli commented on packages/cli/src/models/dependency/Dependency.ts:
```diff
@@ -77,8 +78,16 @@ export default class Dependency {
   public static async install(nameAndVersion: string): Promise<Dependency> {
-    log.info(`Installing ${nameAndVersion} via npm...`);
+    Loggy.add(
+      `${__filename}`,
```
Mmh I forgot to add the `fileName` variable on top of this file as I did in the other ones.

jcarpanelli commented on packages/cli/src/models/Verifier.ts:
```diff
@@ -42,6 +43,7 @@ const Verifier = {
 
 async function publishToEtherchain(
   params: VerifierOptions,
+  remote: string,
```
I added this parameter in order to be able to log the remote name. Probably there's a better way to do it, I'll check it!

jcarpanelli commented on packages/cli/src/models/errors/ErrorHandler.ts:
```diff
@@ -17,9 +19,18 @@ export default class ErrorHandler {
 
   public call(): void {
     if (!this.verbose) {
+      Loggy.stopAll(SpinnerAction.Fail);
```
Ok, this is interesting. By default, `spinnies#stopAll` will set all those actions as `stopped` and show a `-` as prefix, changing the text color to grey. If you send `fail` as argument, it will set all those actions as failed (i.e., with the cross mark symbol). I can just remove the argument.

jcarpanelli commented on packages/cli/src/models/network/NetworkController.ts:
```diff
@@ -682,18 +716,27 @@ export default class NetworkController {
     }
 
     await this._migrateZosversionIfNeeded();
-    log.info(`Publishing project to ${this.network}...`);
```
I added this message inside the `AppProject.ts` file.

jcarpanelli commented on packages/cli/src/models/local/LocalController.ts:
```diff
@@ -123,10 +126,20 @@ export default class LocalController {
+        `Removing ${contractAlias}`,
+      );
       this.packageFile.unsetContract(contractAlias);
+      Loggy.succeed(`remove-${contractAlias}`);
```
It might be a good idea for the future I think (although I don't like callback functions :stuck_out_tongue:)

jcarpanelli commented:
@spalladino ready for another review! I implemented almost all your suggestions, including the `Loggy` API improvements. 
I also implemented what I missed before your last review:
- Removed all references to the old Logger methods (warn, info, error). 
- Removed all the unused imports, the Logger.silent(true) in some config files 
- Implemented the unpack command to work directly with the new logger instead of using the spinner utils.
- Refactored the `CaptureLogs` class
- Removed/added stdouts as discussed offline.

spalladino commented on packages/lib/src/utils/Logger.ts:
```diff
@@ -132,14 +134,27 @@ export const Loggy: { [key: string]: any } = {
+  _actionToText(action: SpinnerAction): string {
+    switch (action) {
+      case SpinnerAction.Add || SpinnerAction.NonSpinnable:
+        return 'added';
```
I'd change this to 'started' (I cannot add suggestions reviewing commit-by-commit)

spalladino commented on packages/lib/src/utils/Logger.ts:
```diff
@@ -132,14 +134,27 @@ export const Loggy: { [key: string]: any } = {
-      case LogType.Success:
-        return 'blueBright';
+      case LogType.Err:
+        return 'redBright';
```
What happened with LogType.Success? And Fail?

jcarpanelli replied:
I removed them as we were not using them.
